twenty-first/src/util_types/merkle_tree.rs

```diff
@@ -1033,4 +1042,166 @@ pub mod merkle_tree_test {
 
         assert_eq!(expected_paths, auth_paths);
     }
+
+    fn root_from_authentication_struct(
+        auth_struct: Vec<Digest>,
+        auth_struct_indices: Vec<u64>,
+        leafs: Vec<Digest>,
+        leafs_node_indices: Vec<u64>,
+        nd_siblings: Vec<(Digest, Digest)>,
+        nd_indices: Vec<(u64, u64)>,
+    ) -> Digest {
+        fn digest_to_xfe(digest: Digest, challenge: XFieldElement) -> XFieldElement {
+            let leaf_xfe_lo = XFieldElement::new([digest.0[0], digest.0[1], digest.0[2]]);
+            let leaf_xfe_hi =
+                challenge * XFieldElement::new([digest.0[3], digest.0[4], BFieldElement::one()]);
+
+            leaf_xfe_lo + leaf_xfe_hi
+        }
+
+        assert_eq!(leafs.len(), leafs_node_indices.len());
+        assert_eq!(auth_struct.len(), auth_struct_indices.len());
+        let (alpha, gamma, delta) = {
+            let mut sponge = Tip5::init();
+            sponge.pad_and_absorb_all(&leafs.encode());
+            sponge.pad_and_absorb_all(&auth_struct.encode());
```

Comment:
This is a non-standard mode of use of the sponge construction. While I cannot find an attack, I am concerned that a vulnerability might hide here. If we definitely want this construction (for instance, because it is more efficient to compute in Triton VM) then we need an iron-clad security argument.

The alternative would be something like `sponge.pad_and_absorb_all(&(leafs, auth_struct).encode())`.

Reply:
Your suggestion probably requires allocation in Triton VM. It's preferable if we can avoid that as much as possible.